File: l33tcode/subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        stack: List[int] = []
        result: List[List[int]] = []

        def dfs(pos: int) -> None:
            if pos == len(nums):
                result.append(stack.copy())
                return

            stack.append(nums[pos])
            dfs(pos + 1)
            stack.pop()
            dfs(pos + 1)

        dfs(0)

        return result

    def subsets2(self, nums: List[int]) -> List[List[int]]:
        def sub_rec(pos, prev):
            if pos == len(nums):
                return [[]]

            result = [prev]

            for r_pos in range(pos + 1, len(nums)):
                for s in sub_rec(r_pos, prev + [nums[r_pos]]):
                    result.append(s)

            return result

        return sub_rec(-1, [])

File: l33tcode/subsets.go
package main

func subsets(nums []int) [][]int {
	results := [][]int{}

	results = append(results, []int{})

	for _, num := range nums {
		for pos := range make([]bool, len(results)) {
			new_result := make([]int, len(results[pos]))
			copy(new_result, results[pos])
			new_result = append(new_result, num)
			results = append(results, new_result)
		}
	}

	return results
}
